x-pack/plugins/enterprise_search/public/applications/shared/use_local_storage/use_local_storage.ts

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { useState, useEffect } from 'react';
+
+export const useLocalStorage = <T>(key: string, defaultValue: T) => {
+  const [item, setItem] = useState<T>(getFromStorage());
+
+  function getFromStorage() {
+    const storedItem = window.localStorage.getItem(key);
+    let toStore: T = defaultValue;
+
+    if (storedItem !== null) {
+      try {
+        toStore = JSON.parse(storedItem) as T;
+      } catch (err) {
+        window.localStorage.removeItem(key);
+      }
+    }
```

Comment:
Hm, I'm somewhat confused by the role of `defaultValue` in this helper. I see it being used only here in which means it's being stored in state and returns to any component using it, but it's not actually used ever when _setting_ local storage, which means if someone initializes `useLocalStorage('foo', 'bar')` but doesn't do anything else, they might expect `foo` to be set to `bar` in their local storage & inspect/look for it, only to find it's not there.

I'm wonder if we should extend this opinionatedly to:
```suggestion
    let toStore: T = defaultValue;

    if (storedItem !== null) {
      try {
        toStore = JSON.parse(storedItem) as T;
      } catch (err) {
        window.localStorage.removeItem(key);
      }
    } else {
      saveToStorage(toStore);
    }
```

so that we're also creating local storage on init - what do you think? 🤷

Reply:
I don't know. I'll write a test for the default value scenario and see what it looks like, then decide.

Reply:
I added a comment to this hook: b34ce9002e30abba75995ccbe289de20ba2ce1c8

```
/**
 * A hook that works like `useState`, but persisted to localStorage.
 *
 * example:
 *
 * const [foo, setFoo] = useLocalStorage("foo", "bar");
 *
 * console.log(foo) // "bar"
 * setFoo("baz")
 * console.log(foo) // "baz"
 *
 * // Navigate away from page and return
 *
 * const [foo, setFoo] = useLocalStorage("foo", "bar");
 * console.log(foo) // "baz"
 */
```

Hopefully that helps clarify the role of defaultValue.

Reply:
I added a comment to this hook: b34ce9002e30abba75995ccbe289de20ba2ce1c8

```
/**
 * A hook that works like `useState`, but persisted to localStorage.
 *
 * example:
 *
 * const [foo, setFoo] = useLocalStorage("foo", "bar");
 *
 * console.log(foo) // "bar"
 * setFoo("baz")
 * console.log(foo) // "baz"
 *
 * // Navigate away from page and return
 *
 * const [foo, setFoo] = useLocalStorage("foo", "bar");
 * console.log(foo) // "baz"
 */
```

Hopefully that helps clarify the role of defaultValue.

Reply:
Just because useLocalStorage **returns** the defaultValue and stores it in **internal** React state does not mean it has actually updated `window.localStorage` itself with defaultValue. That's my concern with the original code - it's not treating `window.localStorage` as the canonical source of truth because it's somewhat conflating its internal state with that. Let me know if you want to hop on a call to clarify?

Reply:
Ah, ok. I see your point now.

Yeah, I'm a little on the fence about this. I'll just make the change, I don't think it would hurt.

Reply:
Sounds good, thanks! Also I just thought of this, an example problem that this might solve is multiple instances of `useLocalStorage` for the same key:

```
const ComponentA = () => {
  useLocalStorage('foo', 'bar');
}
const ComponentB = () => {
  useLocalStorage('foo', 'baz');
}

<ComponentA /> // foo is now bar
<ComponentB /> // foo is still bar, would have been baz without the window.localStorage change
```

It's a bit contrived, but hopefully helps illustrate the point 😅

Reply:
Ugh, that gets hairy, doesn't it 😫.

Good point, definitely agree with you now.

Reply:
7d3ec09eb5a635274c79015d71ae09e009399ccf